questionpy_server/cache.py

```diff
@@ -0,0 +1,152 @@
+
+        return self._files[key]
+
+    def get(self, key: str) -> Path:
+        return self._get_file(key).path
+
+    def remove(self, key: str) -> None:
+        file = self._get_file(key)
+
+        file.path.unlink(missing_ok=True)
+        self._total_bytes -= file.size
+
+        del self._files[key]
+
+    def clear(self) -> None:
+        for file in self._files.values():
+            file.path.unlink(missing_ok=True)
+
+        self._total_bytes = 0
+        self._files.clear()
+
+    def put(self, key: str, value: bytes) -> Path:
+        """
+        Only accepts `bytes` objects as values; raises a `TypeError` otherwise.
+        If the length/size of the provided `value` exceeds `.max_bytes` a `SizeError` is raised.
+        The internal `._total_bytes` attribute is updated.
+        If the key existed before and just the value is replaced, the item is treated as most recently accessed and
+        thus moved to the end of the internal linked list.
+        If after adding the item `._total_bytes` exceeds `.max_bytes`, items are deleted in order from least to most
+        recently accessed until the total size (in bytes) is in line with the specified maximum.
+        """
+
+        if not isinstance(value, bytes):
+            raise TypeError("Not a bytes object:", repr(value))
+
+        size = len(value)
+        if size > self.max_bytes:
+            # If we allowed this, the loop at the end would remove all items from the dictionary,
+            # so we raise an error to allow exceptions for this case.
+            raise SizeError(f"Item itself exceeds maximum allowed size of {self.max_bytes} bytes",
+                            max_size=self.max_bytes, actual_size=size)
+
+        # Update `_total_bytes` depending on whether the key existed already or not.
+        self._total_bytes += size - self._get_file(key).size if key in self._files else size
+
+        # Save the bytes on filesystem.
+        path = self._path(key)
+        assert size == path.open('wb').write(value)
+
+        # Update internal file dictionary.
+        self._files[key] = File(path, size)
+
+        # If size is too large now, remove items until it is less than or equal to the defined maximum.
+        while self._total_bytes > self.max_bytes:
+            # Delete the current oldest item, by instantiating an iterator over all keys (in order)
+            # and passing its next item (i.e. the first one in order) to self.remove.
+            self.remove(next(iter(self._files)))
+
+        return path
+
+    @property
+    def total_bytes(self) -> int:
+        return self._total_bytes
+
+    @property
+    def space_left(self) -> int:
+        return self.max_bytes - self._total_bytes
+
+    @property
+    def directory(self) -> Path:
+        # Ensures that the cache directory exists.
+        if not self._directory.is_dir():
+            self._directory.mkdir(parents=True, exist_ok=True)
+        return self._directory
```

Comment:
Das Verzeichnis wird doch schon in `__init__` erstellt? Falls das Ziel ist, es nach potenzieller Löschung wieder zu erstellen bevor etwas hinzugefügt wird, würde ich das eher in `put` sehen und nicht in einer Property versteckt. Zumal diese Property eh nur innerhalb der Klasse benutzt wird.

Reply:
Genau, nach einer Löschung sollte der Ordner wieder erstellt werden. Wie du schon erwähnt hast, wird die Funktion nur in `put` benötigt, weshalb das Auslagern der Logik in die Property nicht nötig ist - ich habe das angepasst.